File: app/iosApp/ListenUp/Core/Push/PushForegroundPolicy.swift
import Foundation

/// Decides whether an arriving push should show a system banner while the app is in the foreground.
///
/// Pure and unit-tested rather than inline in `PushCoordinator`, because the rule has a genuine
/// exception and getting it wrong is invisible: a suppressed banner looks exactly like a push that
/// never arrived. Mirrors `PlayerGestureMath` — and the Android side's `PushForegroundPolicy`, so
/// the two platforms answer this the same way.
enum PushForegroundPolicy {
    /// Whether the payload in `userInfo` should present while the app is foregrounded.
    ///
    /// A visible app normally suppresses the banner — the running UI is the better surface. A test
    /// notification is the exception, and has to be: it exists purely to prove a notification can
    /// reach this device, and it is triggered from Settings, so the app is *necessarily* foregrounded
    /// when it lands. Suppressing it guaranteed the one thing it was built to demonstrate could
    /// never be demonstrated.
    static func presentsInForeground(userInfo: [AnyHashable: Any]) -> Bool {
        isTestNotification(userInfo: userInfo)
    }

    /// `true` when the relay's opaque payload is a test notification.
    ///
    /// Reads the payload the same way the notification-service extension does — the relay forwards
    /// it as a JSON *string* under `payload`. Deliberately duplicated rather than shared with
    /// `PushPayloadContent`: that type lives in the extension, which does not link the Shared
    /// framework so it can start within its few-second budget.
    static func isTestNotification(userInfo: [AnyHashable: Any]) -> Bool {
        guard let raw = userInfo["payload"] as? String,
              let data = raw.data(using: .utf8),
              let object = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
              let type = object["type"] as? String
        else { return false }
        return type == "test"
    }
}

File: app/sharedUI/src/androidMain/kotlin/com/calypsan/listenup/client/push/PushForegroundPolicy.kt
package com.calypsan.listenup.client.push

import com.calypsan.listenup.api.push.PushPayload

/**
 * Decides whether an arriving push should be rendered as a system notification, given whether the
 * app is currently in the foreground.
 *
 * Pure and unit-tested rather than inline in [ListenUpMessagingService], because the rule has a
 * genuine exception and getting it wrong is invisible: a suppressed notification looks exactly like
 * one that never arrived. Mirrors the `PlayerGestureMath` / `NowPlayingGestureMath` precedent.
 */
internal object PushForegroundPolicy {
    /**
     * Whether [payload] should be rendered.
     *
     * A foregrounded app normally skips the local notification — the in-app notification inbox
     * (bell + list) already shows the row, so the badge has moved by the time the banner would
     * render. [PushPayload.TestNotification] is the exception,
     * and has to be: it exists purely to prove a notification can reach this device, and it is
     * triggered from Settings — which means the app is necessarily in the foreground when it
     * arrives. Suppressing it guaranteed the one thing it was built to demonstrate could never be
     * demonstrated, while the UI cheerfully reported "sent".
     */
    fun shouldRender(
        payload: PushPayload?,
        appInForeground: Boolean,
    ): Boolean = payload is PushPayload.TestNotification || !appInForeground
}
